Smaller bundles, new Rollup scripts, SVG support in JSX, and Typescript/JSDoc fixes

heyheyhello opened the pull request:
Developers using JSX currently can't use SVGs with Sinuous, because transpilation only supports one hyperscript pragma. Sinuous has two, `h` and `hs`, in `` html`...` `` and `` svg`...` `` respectively.

Sidenote: I know JSX is supported as a view syntax, but there doesn't seem to be a lot of detailed examples of it. I found HTM to be rather fragile (#75 got me for an hour) and lead to a not great developer experience. I went with JSX via "@babel/plugin-transform-react-jsx" - I don't agree with the readme's recommendation to transpile JSX to HTM and then HTM to `h`. I'll try to put up a PR soon on how I'm doing it.

Here's how I'm working around the issue:

![image](https://user-images.githubusercontent.com/44614862/82529787-9cb29800-9af0-11ea-87db-70f217f46b48.png)

Let me know your thoughts. Seems to work well, but I understand that a minority of users who actually make use of it, if most are using `html`/`svg` - I know you run a tight ship on bundle size, so it might be better as some kind of module instead of being in _packages/sinuous/src/index.js_

heyheyhello commented on packages/sinuous/src/index.d.ts:
```diff
@@ -34,6 +34,8 @@ declare namespace sinuous {
   const html: (strings: TemplateStringsArray, ...values: any[]) => HTMLElement | DocumentFragment;
   const svg: (strings: TemplateStringsArray, ...values: any[]) => SVGElement | DocumentFragment;
 
+  const svgJSX: <T extends () => unknown>(closure: T) => ReturnType<T>;
```
It's typed ✨ Elements hover as `h.JSX.Element` like every other component

heyheyhello commented on packages/sinuous/src/index.d.ts:
```diff
@@ -92,10 +94,10 @@ declare namespace sinuous {
-   * @param options
-   * @param isSvg
+   * @param apiOverloads Written into the API object
+   * @param state For inflight changes; `hs` is fixed to `{ svgMode: true }`
```
I'm kinda implying in this sentence that state could be used for more than this, which may not be true if the project is stabilizing. I just needed an object reference.

luwes commented:
thanks for your contributions! good points, I never tried SVG in JSX, you're right that would not work with one pragma.

yea the additions would add too much bytes to the core I think.
I wonder if it's possible to make the babel tooling smarter and switch pragma based on tag name or being in a svg tag. that would be ideal

heyheyhello commented:
I'm still looking into it, and I don't know if the workaround right now is the right solution. Maybe leave this open for a bit and we'll see if anyone has ideas

May 23, 2020, 8:47 AM by notifications@github.com:

>
>
>
>
> thanks for your contributions! good points, I never tried SVG in JSX, you're right that would not work with one pragma.
>
>
>
> yea the additions would add too much bytes to the core I think.
>  I wonder if it's possible to make the babel tooling smarter and switch pragma based on tag name or being in a svg tag. that would be ideal
>
>
>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, > view it on GitHub <https://github.com/luwes/sinuous/pull/112#issuecomment-633080069>> , or > unsubscribe <https://github.com/notifications/unsubscribe-auth/AKUMJTU2NU7ITVFJSGAXIODRS7VY3ANCNFSM4NGSY5BA>> .
>
>
>

heyheyhello commented:
Ok I ended up going down a very deep rabbit hole... **Most of the work I'm covering in this post is not related to `svgJSX()`**... Not sure if I should make a new PR or 4-5 PRs.

I was able to **reduce all bundle sizes by up to -44B (according to your size plugin) while still supporting `svgJSX()`**. It's a bit of witchcraft and I'm expecting some hesitation - for instance I haven't decided on the right way to structure the Rollup scripts so I prepared alternatives to open some discussion. Also I removed things that you might want - like `context()`. So. Let me know if that's OK.

The branch is at https://github.com/heyheyhello/sinuous/tree/master and is 27 commits ahead. Here's a summary in reverse commit order (basically reading from the [commits](https://github.com/heyheyhello/sinuous/commits/master)):

**Summary of changes**

- Expose all `api` methods and use them internally within sinuous/h so developers can customize everything. I needed this because I'm tapping into `api.add()` to track when components mount but you're importing and calling `add()` directly instead.
- Introduce `api.s` (boolean-like) to toggle SVG-mode and replace the `context(..., isSvg)` in sinuous/h
- Support SVG in JSX via the `svgJSX()` wrapper which uses `api.s` https://github.com/heyheyhello/sinuous/commit/75bf5e4cb2a3407676e9a6a183f760cde43c1e6f#diff-ecbb6204ac9f66503644e5677ae68712R35-R37
- New sinuous/jsx entrypoint bundle that doesn't import HTM.
  - A bundler might be able to tree-shake this but I use Snowpack and native ES modules in the browser (no bundler) so I end up pulling in HTM when I never use it.
- Type everything with JSDocs so Typescript can work (I use checkJS in VSCode) ✨
- Refactor Typescript _index.d.ts_ files and introduce one for sinuous/h ✨
- Use arrow functions everywhere for smaller bundles. Only applied to ESM in the end since Babel transpiles them out.
- Rewrote all Rollup scripts - I was distracted, more on this below.
  - It's all typed now via JSDoc so your editor will autocomplete Rollup configuration options.
  - Wrote a custom `replace()` plugin that uses MagicString to regex replace out parts of the minified bundle. This is great since Babel doesn't always do the best thing so we can make micro optimizations post-Terser. https://github.com/heyheyhello/sinuous/commit/c911f13e01a42a4be66f1ef200bf0346752cca20
  - Fixed the `npm run build -- --filterName <> --filterType <>` arguments. They weren't working before. Also `--name` was conflicting with Rollup https://github.com/heyheyhello/sinuous/commit/2336caeb6a8e40e9bed7cf86f7215b9e2a921a2b

---

I actually started rewriting the Rollup configs because I wanted to be able to do those regex-replace micro optimizations per-bundle-format-per-entrypoint and the way you wrote it only had one config per entrypoint. However, as I worked on it and eventually did all the replacements I wanted to do, I realized they'll general enough to apply to all entrypoints... So the [scripts/bundles.js](https://github.com/heyheyhello/sinuous/blob/master/scripts/bundles.js) has all the `[ESM]: {}` blocks empty :sweat_smile: All replacements happen in the "global" area afterwards.

This put me on the fence as to whether the per-format configs were worth it, [since `mk()` is a bit complicated and does object-merging](https://github.com/heyheyhello/sinuous/blob/master/scripts/bundles.js#L142), so I made some alternatives in https://github.com/heyheyhello/sinuous/tree/master-alt if you have any feedback. I put pro/cons in the commits.

- Alternative 1: Less complex `mk()` but loses the ability to do per-format configs: https://github.com/heyheyhello/sinuous/commit/ad311d4df4ae2eb321c2602b41b06046d4c200a5#commitcomment-39593479
- Alternative 2: Least magic. Harder to read? Breaks `npm run build` bundle name filtering: https://github.com/heyheyhello/sinuous/commit/94259f6e1a4e89af415be77d924ce24ca9eabd08#commitcomment-39593282

---

Anyway here are the bundle sizes compared to current master:

```
today@twothirty ~/_/w/s/sinuous (master-rebase-alt)> npm run build

> sinuous@0.0.0 build /home/today/_/work/stayknit/sinuous
> rollup -c scripts/rollup.js --silent

                    htm.js ESM  ⏤  847 B (-8 B)
                    htm.js UMD  ⏤  998 B
                htm.min.js IIFE ⏤  937 B
                hydrate.js ESM  ⏤  1.04 kB
                hydrate.js UMD  ⏤  1.1 kB
            hydrate.min.js IIFE ⏤  1.01 kB
             observable.js ESM  ⏤  707 B (-8 B)
             observable.js UMD  ⏤  782 B
         observable.min.js IIFE ⏤  704 B
                      h.js ESM  ⏤  985 B (-44 B)
                      h.js UMD  ⏤  1.07 kB (-2 B)
                  h.min.js IIFE ⏤  1 kB (-5 B)
               template.js ESM  ⏤  710 B (-4 B)
               template.js UMD  ⏤  798 B
           template.min.js IIFE ⏤  713 B
                   data.js ESM  ⏤  503 B (-5 B)
                   data.js UMD  ⏤  587 B
               data.min.js IIFE ⏤  499 B
                   memo.js ESM  ⏤  374 B (-5 B)
                   memo.js UMD  ⏤  438 B
               memo.min.js IIFE ⏤  372 B
                 render.js ESM  ⏤  445 B
                 render.js UMD  ⏤  529 B
             render.min.js IIFE ⏤  423 B
                   mini.js ESM  ⏤  755 B (-3 B)
                   mini.js UMD  ⏤  827 B
               mini.min.js IIFE ⏤  739 B
                    map.js ESM  ⏤  902 B (-2 B)
                    map.js UMD  ⏤  969 B
                map.min.js IIFE ⏤  881 B
                sinuous.js ESM  ⏤  1.14 kB (-16 B)
                sinuous.js UMD  ⏤  1.23 kB (-17 B)
            sinuous.min.js IIFE ⏤  1.13 kB (-22 B)
     sinuous-observable.js ESM  ⏤  1.44 kB (-28 B)
                    jsx.js ESM  ⏤  1.09 kB (+1.09 kB)
                    jsx.js UMD  ⏤  1.18 kB (+1.18 kB)
                jsx.min.js IIFE ⏤  1.1 kB (+1.1 kB)
       babel-plugin-htm.js ESM  ⏤  2.08 kB (-12 B)
       babel-plugin-htm.js CJS  ⏤  4.2 kB
                    all.js ESM  ⏤  3.52 kB (-26 B)
                    all.js UMD  ⏤  3.68 kB (+2 B)
                all.min.js IIFE ⏤  3.61 kB
```

Let me know what you think and what parts of this you'd like to see PRs for.

heyheyhello commented:
I'd also be happy to have `svgJSX()` and sinuous/jsx rejected as long as `isSvg` in sinuous/h could consider `api.s` and use it like `api.insert(..., isSvg || api.s)`. That way we could document to those using JSX that they should write their own version of sinuous/jsx outside of the framework in their own code

heyheyhello commented:
Force pushed since no amount of links I provide in my comments will beat a diff/compare

luwes commented:
@heyheyhello wow, thank you for putting so much effort in this.

yes, it would be easier to review if each part was isolated in a PR.

these are great additions!

> - Expose all api methods and use them internally within sinuous/h so developers can customize everything. I needed this because I'm tapping into api.add() to track when components mount but you're importing and calling add() directly instead.
> - Introduce api.s (boolean-like) to toggle SVG-mode and replace the context(..., isSvg) in sinuous/h
> - Support SVG in JSX via the svgJSX() wrapper which uses api.s heyheyhello@75bf5e4#diff-ecbb6204ac9f66503644e5677ae68712R35-R37
> - New sinuous/jsx entrypoint bundle that doesn't import HTM.
A bundler might be able to tree-shake this but I use Snowpack and native ES modules in the browser (no bundler) so I end up pulling in HTM when I never use it.
> - Type everything with JSDocs so Typescript can work (I use checkJS in VSCode) ✨
> - Refactor Typescript index.d.ts files and introduce one for sinuous/h ✨
> - Use arrow functions everywhere for smaller bundles. Only applied to ESM in the end since Babel transpiles them out.
> - Rewrote all Rollup scripts - I was distracted, more on this below.
It's all typed now via JSDoc so your editor will autocomplete Rollup configuration options.


some things I would prefer to keep how it is:

- the build paths `/module`, `/dist`
	looks cleaner, it is less informative I guess but there's just 2 options umd or esm
- the build scripts I'm not so sure about everything yet, I'll have a deeper look soon.
- I'll have to check the api.s function more but `context` is for https://github.com/luwes/sinuous/wiki/Choose-your-own-reactive-library
	the examples on the wiki are broken by the way but it would use `context`

heyheyhello commented:
Right, I see how `context()` is used in https://github.com/luwes/sinuous/wiki/Choose-your-own-reactive-library but I think that could just be written as:

```diff
- import sinuous from 'sinuous';
+ import { api, h } from 'sinuous';
import htm from 'htm';
import hyperactiv from 'hyperactiv';
const { observe, computed } = hyperactiv;

- const h = sinuous({ subscribe: computed });
+ api.subscribe = computed;
const html = htm.bind(h);
const randomColor = () => '#' + ((Math.random() * (1 << 24)) | 0).toString(16);

const s = observe({
  count: 0,
  style: ''
});

const view = () => {
  return html`
    <h1 style=${() => s.style}>Sinuous <sup>${() => s.count}</sup></h1>
  `;
};

document.querySelector('.sinuous').append(view());
setInterval(() => (s.style = { color: randomColor() }) && s.count++, 1000);
``` 

So `h()` will use Hyperactiv's computed via API

heyheyhello commented:
You'll have to update the Node version on Netlify for the builds to support Array.flat()

luwes commented:
I'm gonna start cherry-picking some of your commits, add them to master.
great work!

heyheyhello commented:
Hey sorry I didn't actually mean to push to this branch, I've been working on the master branch more. I just pushed more type changes right now... (https://github.com/heyheyhello/sinuous/commit/3f07873916fb44d778c53f43c9f973a6e486d727) I could push to this branch too, but I don't wanna overwhelm you this more and more stuff

heyheyhello commented:
If you wanna do a call or chat somewhere I can help explain some of the work or we can talk about breaking it up into more manageable PRs
